### PR TITLE
Allows creating entities 

### PR DESCRIPTION
**Pull request by grafitto**
Adds `project.entities.create` method on @surix/client

### CODE REVIEW

**Comment by habbes** [comment]
@grafitto I see that the user has to specify the data types for each field. Could we abstract that away to make it more developer friendly?

**Comment by habbes** [comment]
Allow the developer to enter  plain objects like
```javascript
entities.create({ name: 'John', active: true })
```
then the library will expand them to
```js
{
  data: {
    name: { type: 'text', value: 'John' },
    active: { type: 'boolean', value: true }
  }
}
```

**Comment by grafitto** [comment]
Didn't think of that. How about the email and phone types? Though I think there are libs for that

**Comment by habbes** [comment]
@grafitto hmm that's a good point. I know there are libs for that, but maybe it's not wise to auto-infer them as phone/email fields without the user explicitly deciding to treat them so. But maybe there's no harm in doing that either.
I think we could have special types for those fields, like a PhoneField('+254....'), and then lib can check whether the field is an instance of that type.
But for the purpose of the MVP, maybe we can skip phone/email support. What do you think?

**Comment by grafitto** [comment]
We can skip that for now, meanwhile we can brainstorm on how it should be implemented in future

**Comment by habbes** [comment]
The reason I think it's important to have the conversion, is to keep the API simple, especially since we don't have an easy way for the MVP users to import data in batch, we should at least make it really for them to write an import script that's easy as just passing a plain JSON object to the library method without writing a lot of code to convert their plain objects to Surix's entities.

**Comment by grafitto** [comment]
Also how do we allow one to put `tags` when creating entities?

**Comment by habbes** [comment]
@grafitto, that's a good point. Maybe we should just let the user enter data this way:

```javascript
{
  data: { name: 'John'  },
  tags: []
}
```
where `tags` is optional.
What do you think?

**Comment by grafitto** [comment]
> @grafitto, that's a good point. Maybe we should just let the user enter data this way:
> 
> ```js
> {
>   data: { name: 'John'  },
>   tags: []
> }
> ```
> 
> where `tags` is optional.
> What do you think?

I think this is good. I was thinking of having a second parameter in the create function called `tags` but I think that is more complicated.

**Comment by habbes** [comment]
> I think this is good. I was thinking of having a second parameter in the create function called `tags` but I think that is more complicated.

I also had that idea, and honestly, I don't know which is better.

**Comment by habbes** [comment]
But let's stick to the first idea then.

**Comment by grafitto** [comment]
Okay

**Comment by grafitto** [comment]
I will add a `expandEntity` method on @surix/data-helpers then use it in client. What do you think?

**Comment by habbes** [comment]
@grafitto I think that's ideal.

**Comment by grafitto** [comment]
Nice @habbes

**Comment by habbes** [comment]
@grafitto If you'll be expanding just the `data` part of the entity, then I think `expandEntityData` is better. Or `inflateEntityData`. Maybe `expand` sounds better.

**Comment by grafitto** [comment]
@habbes you can now go through the changes

**Comment by habbes** [comment]
@grafitto I've reviewed the code and added some comments. The main issue I've pointed out is that your function does not seem to account for all the field types we'll support in the MVP, i.e. lists, booleans, datetimes

**Comment by habbes** [comment]
@grafitto note that an array like:

```javascript
[1, "two", { foo: 'bar' }, ["four"]
```
should be expanded to:
```javascript
{
 type: 'list',
 value: [
   { type: 'number', value: '1' },
   { type: 'text', value: 'two' },
   { type: 'object', value: { foo: { type: 'text', value: 'bar' } } },
   { type: 'list', value: [ { type: 'text', value: 'four' } ] }
}
```

File: packages/data-helpers/src/expand-entity/expand-entity.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+function expandEntityData (data: DataField): RawEntityData {
+    const expandedEntity = Object.keys(data)
+        .reduce((expanded, key) => {
+        let converted = {};
+        switch(typeof data[key]) {
+            case 'object':
+                const isArray = data[key] instanceof Array;
+                converted = {
+                    type: isArray ? 'list' : 'object',
```
Comment (habbes) [inline packages/data-helpers/src/expand-entity/expand-entity.ts]: I think this block of code (line 17-25) is a bit hard to follow, especially with the `// @ts-ignore` comments interrupting the flow. Maybe it would just been better to use an if statement?

File: packages/data-helpers/src/expand-entity/tests/expand-entity.test.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+            }
+        };
+    });
+    it('should be a function', async () => {
+        expect(expandEntity).toBeInstanceOf(Function);
+    });
+    it('should convert successfully', async () => {
+        const expected = {
+            data: {
+                name: {
```
[the remark this answers is not in the document]
Comment (habbes) [reply packages/data-helpers/src/expand-entity/tests/expand-entity.test.ts]: @grafitto as a suggestion, not necessarily something to implement, you could also check out snapshot testing, it might save you the time of writing this whole `expected` block and having to manually extend it when you add new tests cases, i.e. using `expect(expandEntity(entity)).toMatchSnapshot()`

[the remark this answers is not in the document]
Comment (habbes) [reply packages/data-helpers/src/expand-entity/tests/expand-entity.test.ts]: @grafitto what does `expandEntity` do if the entity has other fields besides `data`, like `_id`, `tags`, etc.?

File: packages/data-helpers/src/expand-entity/expand-entity.ts (hunk 1 above)
Comment (grafitto) [reply packages/data-helpers/src/expand-entity/expand-entity.ts]: I have changed to if statement

File: packages/data-helpers/src/expand-entity/expand-entity.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+                    value: expandEntityData(data[key])
+                };
+            break;
+            case 'string':
+                converted = {
+                    type: 'text',
+                    value: data[key]
+                };
+            break;
+            default:
```
[the remark this answers is not in the document]
Comment (grafitto) [reply packages/data-helpers/src/expand-entity/expand-entity.ts]: Other types are to be handled in the default section. Though I can explicitly add them with case blocks

File: packages/data-helpers/src/expand-entity/tests/expand-entity.test.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+                name: 'Kevin Nderitu',
+                age: 12,
+                address: {
+                    location: 'Nairobi'
+                }
+            }
+        };
+    });
+    it('should be a function', async () => {
+        expect(expandEntity).toBeInstanceOf(Function);
```
[the remark this answers is not in the document]
Comment (grafitto) [reply packages/data-helpers/src/expand-entity/tests/expand-entity.test.ts]: Not sure. I can just remove it if it is not necessary

File: packages/data-helpers/src/expand-entity/tests/expand-entity.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (grafitto) [reply packages/data-helpers/src/expand-entity/tests/expand-entity.test.ts]: Other fields are ignored and returned as they are

File: packages/data-helpers/src/expand-entity/tests/expand-entity.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (grafitto) [reply packages/data-helpers/src/expand-entity/tests/expand-entity.test.ts]: Removed

File: packages/data-helpers/src/expand-entity/tests/expand-entity.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (grafitto) [reply packages/data-helpers/src/expand-entity/tests/expand-entity.test.ts]: @habbes is that what was required?

File: packages/data-helpers/src/expand-entity/expand-entity.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        .reduce((expanded, key) => {
+        let converted = {};
+        switch(typeof data[key]) {
+            case 'object':
+                converted = {
+                    type: 'object',
+                    // @ts-ignore 
+                    value: expandEntityData(data[key])
+                };
+            break;
```
[the remark this answers is not in the document]
Comment (grafitto) [reply packages/data-helpers/src/expand-entity/expand-entity.ts]: Fixed

File: packages/data-helpers/src/expand-entity/expand-entity.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (grafitto) [reply packages/data-helpers/src/expand-entity/expand-entity.ts]: Fixed

File: packages/data-helpers/README.md [hunk 6]
```diff
@@ -37,6 +37,35 @@ const {
 } = require('@surix/data-helpers')
 ```
 
+### `expnadEntity`
```
Comment (habbes) [inline packages/data-helpers/README.md]: There's a typo in the header, it should `expandEntity`

File: packages/data-helpers/src/expand-entity/tests/expand-entity.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (habbes) [reply packages/data-helpers/src/expand-entity/tests/expand-entity.test.ts]: I think that's good, so long as they're not removed
